sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProgressReporter.scala

```diff
@@ -345,7 +345,14 @@ trait ProgressReporter extends Logging {
       val allExecPlanLeaves = lastExecution.executedPlan.collectLeaves()
       if (allLogicalPlanLeaves.size == allExecPlanLeaves.size) {
         val execLeafToSource = allLogicalPlanLeaves.zip(allExecPlanLeaves).flatMap {
-          case (lp, ep) => logicalPlanLeafToSource.get(lp).map { source => ep -> source }
+          case (_, ep: MicroBatchScanExec) =>
+            // SPARK-41199: `logicalPlanLeafToSource` contains OffsetHolder instance for DSv2
+            // streaming source, hence we cannot lookup the actual source from the map.
+            // The physical node for DSv2 streaming source contains the information of the source
+            // by itself, so leverage it.
+            Some(ep -> ep.stream)
```

Comment:
Why it is only for DS v1 mixing with DS v2? Seems DS v2 stream source always cannot be matched by `logicalPlanLeafToSource` because it is `OffsetHolder`.

Reply:
We have two different paths - if there are only DSv2 streaming sources, we don't even try to match the logical plan and physical plan. We just collect the metrics out from physical plan, which is always accurate.

https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProgressReporter.scala#L296-L316

Reply:
So all the problems from streaming query metrics are due to DSv1 which we do not have a dedicated logical node / physical node for source.

Reply:
Oh, got it. I saw that there is a `onlyDataSourceV2Sources` and a dedicated block for it.